File: sdks/dotnet/src/Rrflow.Rrd.Client/EndpointResolver.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Text.RegularExpressions;

namespace Rrflow.Rrd;

internal sealed partial class EndpointResolver
{
    private readonly Uri _baseUri;

    internal EndpointResolver(ClientOptions options)
    {
        _baseUri = options.BaseUri;
        Instance = options.Instance;
    }

    internal string Instance { get; }

    internal Uri Resolve(Endpoint endpoint, IReadOnlyDictionary<string, string> parameters)
    {
        string path = PathParameter().Replace(endpoint.Path, match =>
        {
            string name = match.Groups[1].Value;
            if (!parameters.TryGetValue(name, out string? value) || value.Length == 0)
            {
                throw new RrdClientException($"missing path parameter {name}");
            }
            return Uri.EscapeDataString(CorrelationIdentifier(value, name + " path parameter"));
        });
        return new Uri(_baseUri, path.TrimStart('/'));
    }

    internal static string CorrelationIdentifier(string? value, string label)
    {
        if (value is null
            || value.Length is < 1 or > 128
            || !CorrelationPattern().IsMatch(value))
        {
            throw new RrdClientException($"{label} is not a canonical RRD correlation ID");
        }
        return value;
    }

    internal static string CanonicalIdentifier(string? value, string label)
    {
        if (value is null
            || value.Length is < 1 or > 128
            || !CanonicalPattern().IsMatch(value))
        {
            throw new RrdClientException($"{label} is not a canonical RRD identifier");
        }
        return value;
    }

    internal static Uri RequireLoopbackBaseUri(Uri uri)
    {
        bool loopback = uri.Host.Equals("localhost", StringComparison.OrdinalIgnoreCase)
            || (IPAddress.TryParse(uri.Host, out IPAddress? address) && IPAddress.IsLoopback(address));
        if (uri.Scheme != Uri.UriSchemeHttp
            || !loopback
            || !string.IsNullOrEmpty(uri.UserInfo)
            || !string.IsNullOrEmpty(uri.Query)
            || !string.IsNullOrEmpty(uri.Fragment))
        {
            throw new RrdClientException(
                "RRD .NET client permits only credential-free loopback HTTP before TLS qualification");
        }
        return new Uri(uri.AbsoluteUri.TrimEnd('/') + "/", UriKind.Absolute);
    }

    [GeneratedRegex("^[A-Za-z0-9._:-]+$", RegexOptions.CultureInvariant)]
    private static partial Regex CorrelationPattern();

    [GeneratedRegex("^[a-z0-9][a-z0-9._-]*$", RegexOptions.CultureInvariant)]
    private static partial Regex CanonicalPattern();

    [GeneratedRegex("\\{([a-z]+)\\}", RegexOptions.CultureInvariant)]
    private static partial Regex PathParameter();
}

File: sdks/java/src/main/java/io/rrflow/rrd/EndpointResolver.java
package io.rrflow.rrd;

import java.net.InetAddress;
import java.net.URI;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/** Current loopback endpoint and route resolution; H-07 owns installed identities. */
final class EndpointResolver {
    private static final Pattern PATH_PARAMETER = Pattern.compile("\\{([a-z]+)}");

    private final URI baseUri;

    EndpointResolver(URI baseUri) {
        this.baseUri = baseUri;
    }

    URI resolve(OperationId operation, Map<String, String> parameters) {
        String path = resolvePath(operation.path(), parameters);
        return baseUri.resolve(path.startsWith("/") ? path.substring(1) : path);
    }

    static URI loopbackUri(String value) {
        final URI uri;
        try {
            uri = URI.create(value);
            String host = uri.getHost();
            boolean loopback = "localhost".equals(host);
            if (host != null && !loopback && host.matches("^[0-9a-fA-F:.]+$")) {
                InetAddress address = InetAddress.getByName(host);
                loopback = address.isLoopbackAddress();
            }
            if (!"http".equals(uri.getScheme())
                    || !loopback
                    || uri.getUserInfo() != null
                    || uri.getQuery() != null
                    || uri.getFragment() != null) {
                throw new IllegalArgumentException();
            }
            return URI.create(value.endsWith("/") ? value : value + "/");
        } catch (Exception error) {
            throw new RrdClientException(
                    "RRD Java client permits only credential-free loopback HTTP before TLS qualification",
                    error);
        }
    }

    private static String resolvePath(String template, Map<String, String> parameters) {
        Matcher matcher = PATH_PARAMETER.matcher(template);
        StringBuilder result = new StringBuilder();
        while (matcher.find()) {
            String value = parameters.get(matcher.group(1));
            if (value == null || value.isEmpty()) {
                throw new RrdClientException("missing path parameter " + matcher.group(1));
            }
            String encoded = URLEncoder.encode(
                            OperationBinding.correlation(
                                    value, matcher.group(1) + " path parameter"),
                            StandardCharsets.UTF_8)
                    .replace("+", "%20");
            matcher.appendReplacement(result, Matcher.quoteReplacement(encoded));
        }
        matcher.appendTail(result);
        return result.toString();
    }
}
